File: src/plastic_agent_net/control/controller.py
"""PlasticityController: orchestrate spawn/prune/merge/escalate."""

from __future__ import annotations

import logging
from typing import Any

from plastic_agent_net.config.templates import TEMPLATE_CONFIGS
from plastic_agent_net.control.merge_rules import check_merge_triggers
from plastic_agent_net.control.prune_rules import check_prune_triggers
from plastic_agent_net.control.scoring import compute_branch_score, compute_node_contribution
from plastic_agent_net.control.spawn_rules import check_spawn_triggers
from plastic_agent_net.core.budgets import BudgetWatchdog
from plastic_agent_net.core.graph import AgentGraph
from plastic_agent_net.core.messages import MessageBus
from plastic_agent_net.core.models import (
    ActionType,
    AgentTemplate,
    ControllerAction,
    ControllerState,
    EdgeState,
    ModelTier,
    NodeState,
    NodeStatus,
    PersonaVector,
)
from plastic_agent_net.memory.artifact_store import ArtifactStore

logger = logging.getLogger(__name__)


class PlasticityController:
    """Aggregates signals from rules and applies structural changes to the graph."""

    def __init__(
        self,
        graph: AgentGraph,
        artifact_store: ArtifactStore,
        message_bus: MessageBus,
        budget: BudgetWatchdog,
    ) -> None:
        self._graph = graph
        self._artifact_store = artifact_store
        self._message_bus = message_bus
        self._budget = budget
        self._state = ControllerState()

    def step(self, current_round: int) -> ControllerState:
        """One controller step: aggregate signals → compute actions → apply."""
        self._state = ControllerState(round=current_round)
        pressure = self._budget.pressure()

        # Gather actions from all rule sets
        actions: list[ControllerAction] = []
        actions.extend(check_spawn_triggers(self._graph, self._artifact_store, current_round))
        actions.extend(check_prune_triggers(self._graph, self._artifact_store, current_round, pressure))
        actions.extend(check_merge_triggers(self._graph, self._artifact_store, current_round))
        actions.extend(self._check_escalation(current_round))

        # Filter by budget constraints
        actions = self._filter_by_budget(actions)

        # Apply actions
        for action in actions:
            self._apply(action, current_round)
            self._state.actions_taken.append(action)

        # Update scores
        for bid in self._graph.branch_ids():
            self._state.branch_scores[bid] = compute_branch_score(
                bid, self._graph, self._artifact_store, current_round
            )
        for node in self._graph.active_nodes():
            self._state.node_contributions[node.node_id] = compute_node_contribution(
                node, self._graph, self._artifact_store, current_round
            )

        return self._state

    def _check_escalation(self, current_round: int) -> list[ControllerAction]:
        """Check if any nodes should be escalated to a stronger model."""
        actions = []
        for node in self._graph.active_nodes():
            if node.model_tier == ModelTier.OPUS:
                continue  # Already at max
            if node.rounds_active >= 3 and node.contribution_score < 0.2:
                template_config = TEMPLATE_CONFIGS.get(node.template)
                escalation_tier = template_config.escalation_tier if template_config else ModelTier.SONNET
                actions.append(ControllerAction(
                    action_type=ActionType.ESCALATE,
                    target_node=node.node_id,
                    payload={"new_tier": escalation_tier.value},
                    reason=f"Node {node.node_id} underperforming, escalating to {escalation_tier.value}",
                ))
        return actions

    def _filter_by_budget(self, actions: list[ControllerAction]) -> list[ControllerAction]:
        """Remove actions that would violate budget constraints."""
        filtered = []
        pending_spawns = 0

        for action in actions:
            if action.action_type == ActionType.SPAWN:
                if self._graph.node_count() + pending_spawns >= self._budget.config.max_nodes:
                    logger.info("Skipping spawn: at node limit")
                    continue
                new_branch = action.payload.get("branch_id", "main")
                if new_branch not in self._graph.branch_ids() and len(self._graph.branch_ids()) >= self._budget.config.max_branches:
                    logger.info("Skipping spawn: at branch limit")
                    continue
                pending_spawns += 1

            filtered.append(action)

        return filtered

    def _apply(self, action: ControllerAction, current_round: int) -> None:
        """Apply a single controller action to the graph."""
        if action.action_type == ActionType.SPAWN:
            self._apply_spawn(action, current_round)
        elif action.action_type == ActionType.PRUNE:
            self._apply_prune(action)
        elif action.action_type == ActionType.MERGE:
            self._apply_merge(action)
        elif action.action_type == ActionType.ESCALATE:
            self._apply_escalate(action)
        elif action.action_type == ActionType.REWEIGHT_EDGE:
            self._apply_reweight(action)
        elif action.action_type == ActionType.ADD_EDGE:
            self._apply_add_edge(action)
        elif action.action_type == ActionType.REMOVE_EDGE:
            self._apply_remove_edge(action)

    def _apply_spawn(self, action: ControllerAction, current_round: int) -> None:
        template_str = action.payload.get("template", "coder")
        template = AgentTemplate(template_str)
        branch_id = action.payload.get("branch_id", "main")
        config = TEMPLATE_CONFIGS.get(template)

        node = NodeState(
            template=template,
            persona=config.persona_prior if config else PersonaVector(),
            model_tier=config.default_model_tier if config else ModelTier.HAIKU,
            branch_id=branch_id,
            round_created=current_round,
        )
        self._graph.add_node(node)

        # Connect to existing nodes on same branch
        for existing in self._graph.nodes_by_branch(branch_id):
            if existing.node_id != node.node_id and existing.status != NodeStatus.PRUNED:
                self._graph.add_edge(EdgeState(source=existing.node_id, target=node.node_id))

        logger.info("Spawned %s node %s on branch %s", template.value, node.node_id, branch_id)

    def _apply_prune(self, action: ControllerAction) -> None:
        self._graph.remove_node(action.target_node)
        logger.info("Pruned node %s: %s", action.target_node, action.reason)

    def _apply_merge(self, action: ControllerAction) -> None:
        source_branch = action.payload.get("source_branch", "")
        target_branch = action.payload.get("target_branch", "main")

        for node in self._graph.nodes_by_branch(source_branch):
            node.status = NodeStatus.MERGED
            node.branch_id = target_branch

        logger.info("Merged branch %s into %s", source_branch, target_branch)

    def _apply_escalate(self, action: ControllerAction) -> None:
        node = self._graph.get_node(action.target_node)
        if node:
            new_tier = ModelTier(action.payload.get("new_tier", "sonnet"))
            node.model_tier = new_tier
            logger.info("Escalated node %s to %s", node.node_id, new_tier.value)

    def _apply_reweight(self, action: ControllerAction) -> None:
        source = action.payload.get("source", "")
        target = action.payload.get("target", "")
        weight = action.payload.get("weight", 1.0)
        edge = self._graph.get_edge(source, target)
        if edge:
            edge.weight = weight

    def _apply_add_edge(self, action: ControllerAction) -> None:
        source = action.payload.get("source", "")
        target = action.payload.get("target", "")
        self._graph.add_edge(EdgeState(source=source, target=target))

    def _apply_remove_edge(self, action: ControllerAction) -> None:
        source = action.payload.get("source", "")
        target = action.payload.get("target", "")
        self._graph.remove_edge(source, target)

File: supabase/functions/run-episode/controller.ts
/**
 * Port of control/ modules: scoring, spawn/prune/merge/escalate rules.
 */

import {
  ESCALATION_FAILURE_COUNT,
  MERGE_SIMILARITY_THRESHOLD,
  PRUNE_CONTRIBUTION_THRESHOLD,
  SCORING_WEIGHTS,
  TEMPLATE_CONFIGS,
} from "./templates.ts";

// ============================================================
// Types
// ============================================================

export interface NodeRow {
  id: string;
  template: string;
  branch_id: string;
  status: string;
  round_created: number;
  rounds_active: number;
  tokens_used: number;
  contribution_score: number;
  model_tier: string;
  persona: Record<string, number>;
  metadata: Record<string, unknown>;
}

export interface EdgeRow {
  source_node: string;
  target_node: string;
  weight: number;
  active: boolean;
}

export interface ArtifactRow {
  id: string;
  artifact_type: string;
  producer_node: string;
  branch_id: string;
  round_produced: number;
  content: Record<string, unknown>;
  summary: string;
}

export interface ControllerAction {
  actionType: string;
  targetNode: string;
  payload: Record<string, unknown>;
  reason: string;
}

export interface ControllerState {
  round: number;
  actionsTaken: ControllerAction[];
  branchScores: Record<string, number>;
  nodeContributions: Record<string, number>;
}

// ============================================================
// Scoring
// ============================================================

export function computeNodeContribution(
  node: NodeRow,
  edges: EdgeRow[],
  allNodes: NodeRow[],
  artifacts: ArtifactRow[],
  currentRound: number,
): number {
  const w = SCORING_WEIGHTS;
  const nodeArtifacts = artifacts.filter((a) => a.producer_node === node.id);
  if (nodeArtifacts.length === 0) return 0;

  let score = 0;

  // Novelty
  const typesProduced = new Set(nodeArtifacts.map((a) => a.artifact_type));
  score += w.novelty * Math.min(1.0, typesProduced.size / 3);

  // Verification delta
  const branchVerifs = artifacts
    .filter((a) => a.artifact_type === "verification" && a.branch_id === node.branch_id)
    .sort((a, b) => a.round_produced - b.round_produced);

  if (branchVerifs.length >= 2) {
    const prev = (branchVerifs[branchVerifs.length - 2].content as any).overall_score ?? 0;
    const curr = (branchVerifs[branchVerifs.length - 1].content as any).overall_score ?? 0;
    score += w.verification_delta * Math.max(0, curr - prev);
  } else if (branchVerifs.length === 1) {
    score += w.verification_delta * ((branchVerifs[0].content as any).overall_score ?? 0);
  }

  // Peer agreement
  const successors = edges
    .filter((e) => e.source_node === node.id && e.active)
    .map((e) => e.target_node);
  if (successors.length > 0) {
    const activeSuccessors = successors.filter((s) => {
      const n = allNodes.find((nn) => nn.id === s);
      return n && (n.status === "running" || n.status === "done");
    }).length;
    score += w.peer_agreement * Math.min(1.0, activeSuccessors / Math.max(1, successors.length));
  }

  // Recency
  const age = currentRound - node.round_created;
  score += w.recency * Math.max(0, 1.0 - age / 10);

  return Math.min(1.0, score);
}

export function computeBranchScore(
  branchId: string,
  nodes: NodeRow[],
  edges: EdgeRow[],
  artifacts: ArtifactRow[],
  currentRound: number,
): number {
  const branchVerifs = artifacts
    .filter((a) => a.artifact_type === "verification" && a.branch_id === branchId)
    .sort((a, b) => a.round_produced - b.round_produced);

  let verifScore = 0;
  if (branchVerifs.length > 0) {
    verifScore = (branchVerifs[branchVerifs.length - 1].content as any).overall_score ?? 0;
  }

  const branchNodes = nodes.filter((n) => n.branch_id === branchId);
  let avgContrib = 0;
  if (branchNodes.length > 0) {
    const contribs = branchNodes.map((n) =>
      computeNodeContribution(n, edges, nodes, artifacts, currentRound)
    );
    avgContrib = contribs.reduce((s, v) => s + v, 0) / contribs.length;
  }

  return 0.7 * verifScore + 0.3 * avgContrib;
}

// ============================================================
// Spawn Rules
// ============================================================

function checkSpawnTriggers(
  nodes: NodeRow[],
  edges: EdgeRow[],
  artifacts: ArtifactRow[],
  currentRound: number,
): ControllerAction[] {
  const actions: ControllerAction[] = [];

  // Uncertainty branching
  const plans = artifacts.filter((a) => a.artifact_type === "plan");
  for (const plan of plans) {
    const uncertainties = (plan.content as any).uncertainties ?? [];
    if (uncertainties.length > 0 && (plan.content as any).needs_branching) {
      actions.push({
        actionType: "spawn",
        targetNode: "",
        payload: { template: "coder", branch_id: `branch_${currentRound}` },
        reason: `Plan has ${uncertainties.length} uncertainties, spawning alternative branch`,
      });
    }
  }

  // Repeated failures → spawn debugger
  const verifications = artifacts.filter((a) => a.artifact_type === "verification");
  const branchFailures: Record<string, number> = {};
  for (const v of verifications) {
    const c = v.content as any;
    if (!c.tests_passed || !c.build_passed) {
      branchFailures[v.branch_id] = (branchFailures[v.branch_id] ?? 0) + 1;
    }
  }
  for (const [branchId, count] of Object.entries(branchFailures)) {
    if (count >= ESCALATION_FAILURE_COUNT) {
      const hasDebugger = nodes.some(
        (n) => n.template === "debugger" && n.branch_id === branchId,
      );
      if (!hasDebugger) {
        actions.push({
          actionType: "spawn",
          targetNode: "",
          payload: { template: "debugger", branch_id: branchId },
          reason: `Branch ${branchId} has ${count} verification failures`,
        });
      }
    }
  }

  // Unreviewed patches → spawn reviewer
  const patches = artifacts.filter((a) => a.artifact_type === "patch");
  const reviews = artifacts.filter((a) => a.artifact_type === "review");
  const reviewedBranches = new Set(reviews.map((r) => r.branch_id));
  for (const patch of patches) {
    if (!reviewedBranches.has(patch.branch_id)) {
      const hasReviewer = nodes.some(
        (n) =>
          (n.template === "skeptic_reviewer" || n.template === "security_reviewer") &&
          n.branch_id === patch.branch_id,
      );
      if (!hasReviewer) {
        actions.push({
          actionType: "spawn",
          targetNode: "",
          payload: { template: "skeptic_reviewer", branch_id: patch.branch_id },
          reason: `Patch on branch ${patch.branch_id} has no review`,
        });
      }
    }
  }

  return actions;
}

// ============================================================
// Prune Rules
// ============================================================

function checkPruneTriggers(
  nodes: NodeRow[],
  edges: EdgeRow[],
  artifacts: ArtifactRow[],
  currentRound: number,
  budgetPressure: number,
): ControllerAction[] {
  const actions: ControllerAction[] = [];
  const activeNodes = nodes.filter((n) => n.status === "pending" || n.status === "running");

  for (const node of activeNodes) {
    if (node.rounds_active < 2) continue;
    const contrib = computeNodeContribution(node, edges, nodes, artifacts, currentRound);
    let threshold = PRUNE_CONTRIBUTION_THRESHOLD;
    if (budgetPressure > 0.7) threshold *= 2;
    if (contrib < threshold) {
      actions.push({
        actionType: "prune",
        targetNode: node.id,
        payload: {},
        reason: `Low contribution (${contrib.toFixed(2)} < ${threshold.toFixed(2)})`,
      });
    }
  }

  // Prune dominated branches
  const branchIds = [...new Set(activeNodes.map((n) => n.branch_id))];
  if (branchIds.length > 1) {
    const branchScores: Record<string, number> = {};
    for (const bid of branchIds) {
      branchScores[bid] = computeBranchScore(bid, nodes, edges, artifacts, currentRound);
    }
    const bestScore = Math.max(...Object.values(branchScores));
    for (const [bid, score] of Object.entries(branchScores)) {
      if (bid === "main") continue;
      if (score < bestScore * 0.3 && currentRound > 3) {
        for (const node of nodes.filter((n) => n.branch_id === bid)) {
          if (node.status === "pending" || node.status === "running") {
            actions.push({
              actionType: "prune",
              targetNode: node.id,
              payload: {},
              reason: `Branch ${bid} dominated (score ${score.toFixed(2)} vs best ${bestScore.toFixed(2)})`,
            });
          }
        }
      }
    }
  }

  return actions;
}

// ============================================================
// Merge Rules
// ============================================================

function checkMergeTriggers(
  nodes: NodeRow[],
  edges: EdgeRow[],
  artifacts: ArtifactRow[],
  currentRound: number,
): ControllerAction[] {
  const actions: ControllerAction[] = [];
  const activeNodes = nodes.filter(
    (n) => n.status !== "pruned" && n.status !== "merged",
  );
  const branchIds = [...new Set(activeNodes.map((n) => n.branch_id))];
  if (branchIds.length < 2) return actions;

  // Near-duplicate branches
  const branchPatches: Record<string, Set<string>> = {};
  for (const bid of branchIds) {
    const patches = artifacts.filter(
      (a) => a.artifact_type === "patch" && a.branch_id === bid,
    );
    const files = new Set<string>();
    for (const p of patches) {
      for (const patch of (p.content as any).patches ?? []) {
        files.add(patch.file_path ?? "");
      }
    }
    branchPatches[bid] = files;
  }

  const checked = new Set<string>();
  for (const b1 of branchIds) {
    for (const b2 of branchIds) {
      const key = [b1, b2].sort().join("|");
      if (b1 === b2 || checked.has(key)) continue;
      checked.add(key);

      const f1 = branchPatches[b1] ?? new Set();
      const f2 = branchPatches[b2] ?? new Set();
      if (f1.size === 0 || f2.size === 0) continue;

      const union = new Set([...f1, ...f2]);
      const intersection = [...f1].filter((x) => f2.has(x));
      const overlap = intersection.length / Math.max(1, union.size);

      if (overlap >= MERGE_SIMILARITY_THRESHOLD) {
        const s1 = computeBranchScore(b1, nodes, edges, artifacts, currentRound);
        const s2 = computeBranchScore(b2, nodes, edges, artifacts, currentRound);
        const [winner, loser] = s1 >= s2 ? [b1, b2] : [b2, b1];
        actions.push({
          actionType: "merge",
          targetNode: "",
          payload: { source_branch: loser, target_branch: winner, overlap },
          reason: `Branches ${loser}→${winner} overlap ${(overlap * 100).toFixed(0)}%`,
        });
      }
    }
  }

  // Converged branches
  for (const bid of branchIds) {
    if (bid === "main") continue;
    const score = computeBranchScore(bid, nodes, edges, artifacts, currentRound);
    const mainScore = computeBranchScore("main", nodes, edges, artifacts, currentRound);
    if (score > 0.8 && mainScore > 0.8 && currentRound > 5) {
      actions.push({
        actionType: "merge",
        targetNode: "",
        payload: { source_branch: bid, target_branch: "main", convergence: true },
        reason: `Branch ${bid} converged with main (both > 0.8)`,
      });
    }
  }

  return actions;
}

// ============================================================
// Escalation
// ============================================================

function checkEscalation(nodes: NodeRow[]): ControllerAction[] {
  const actions: ControllerAction[] = [];
  const activeNodes = nodes.filter((n) => n.status === "pending" || n.status === "running");

  for (const node of activeNodes) {
    if (node.model_tier === "opus") continue;
    if (node.rounds_active >= 3 && node.contribution_score < 0.2) {
      const cfg = TEMPLATE_CONFIGS[node.template];
      const newTier = cfg?.escalationTier ?? "sonnet";
      actions.push({
        actionType: "escalate",
        targetNode: node.id,
        payload: { new_tier: newTier },
        reason: `Node ${node.id} underperforming, escalating to ${newTier}`,
      });
    }
  }

  return actions;
}

// ============================================================
// Main Controller Step
// ============================================================

export function controllerStep(
  nodes: NodeRow[],
  edges: EdgeRow[],
  artifacts: ArtifactRow[],
  currentRound: number,
  budgetPressure: number,
  maxNodes: number,
  maxBranches: number,
): ControllerState {
  let actions: ControllerAction[] = [];

  actions.push(...checkSpawnTriggers(nodes, edges, artifacts, currentRound));
  actions.push(...checkPruneTriggers(nodes, edges, artifacts, currentRound, budgetPressure));
  actions.push(...checkMergeTriggers(nodes, edges, artifacts, currentRound));
  actions.push(...checkEscalation(nodes));

  // Filter spawns by budget
  const activeNodes = nodes.filter((n) => n.status === "pending" || n.status === "running");
  const activeBranches = new Set(
    nodes.filter((n) => n.status !== "pruned" && n.status !== "merged").map((n) => n.branch_id),
  );
  let pendingSpawns = 0;

  actions = actions.filter((a) => {
    if (a.actionType === "spawn") {
      if (activeNodes.length + pendingSpawns >= maxNodes) return false;
      const newBranch = (a.payload.branch_id as string) ?? "main";
      if (!activeBranches.has(newBranch) && activeBranches.size >= maxBranches) return false;
      pendingSpawns++;
    }
    return true;
  });

  // Compute scores
  const branchIds = [...new Set(
    nodes.filter((n) => n.status !== "pruned" && n.status !== "merged").map((n) => n.branch_id),
  )];
  const branchScores: Record<string, number> = {};
  for (const bid of branchIds) {
    branchScores[bid] = computeBranchScore(bid, nodes, edges, artifacts, currentRound);
  }

  const nodeContributions: Record<string, number> = {};
  for (const node of activeNodes) {
    nodeContributions[node.id] = computeNodeContribution(
      node, edges, nodes, artifacts, currentRound,
    );
  }

  return {
    round: currentRound,
    actionsTaken: actions,
    branchScores,
    nodeContributions,
  };
}
